### PR TITLE
Feature/reset password

### PR DESCRIPTION
**Pull request by RobertWSaunders**
*Fixes ticket #18 *

### Changes proposed in this PR:
- adds rudimentary reset password logic in, still needs error messages, form validation, styling and email templates, will be adding in another PR
- adds reset password emails

@zackharley can you look at `mailer/index.js` and tell me a good solution for the url in the email, it changes depending on production/development, is there a way you know of to make it dynamic without add a `HOST` environment variable?

### CODE REVIEW

**Comment by RobertWSaunders** [comment]
Fixing somethings up in this right now!

**Comment by RobertWSaunders** [comment]
Should be good for a review now @zackharley

File: package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
     "redux-saga": "^0.16.0",
     "semantic-ui-less": "^2.2.12",
     "semantic-ui-react": "^0.76.0",
+    "sendgrid": "^5.2.3",
```
Comment (zackharley) [inline package.json]: I think you might want [@sengrid/mail](https://www.npmjs.com/package/@sendgrid/mail) instead.

File: server/mailer/index.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
+const sendgrid = require('@sendgrid/mail');
+
+const { SENDGRID_API_KEY, EMAIL_URL_HOST } = process.env;
+
+sendgrid.setApiKey(SENDGRID_API_KEY);
+
+const QHACKS_NO_REPLY_EMAIL = `no-reply@qhacks.io`;
```
Comment (10kc-zack) [inline server/mailer/index.js]: Should this be using backticks?

File: server/mailer/index.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+const ERROR_MESSAGES = {
+	RESET_PASSWORD_NO_SEND: "The reset password email did not send correctly!",
+	SUCCESSFUL_RESET_NO_SEND: "The successful reset password email did not send correctly!",
+	SUCCESSFUL_APPLICATION_NO_SEND: "The successful application email did not send correctly!"
+};
+
+/**
+ * Creates a standard format error, see routes for usage.
```
Comment (10kc-zack) [inline server/mailer/index.js]: Incorrect comment. Shouldn't be added if it is incorrect.

File: server/mailer/index.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * Sends an email to the user when they apply!
+ * @param {Object} user The user who has applied.
+ * @return {Promise} SendGrid request promise.
+ */
+function sendSuccessfulApplicationEmail(user) {
+	return new Promise((resolve, reject) => {
+		if (!SENDGRID_API_KEY) resolve();
+
+		const request = createEmailMsg({ toEmail: user.email, subject: 'QHacks Application Success!', textContent: 'Thank you for applying!', htmlContent: '<p>Thank you for applying!</p>' });
```
Comment (10kc-zack) [inline server/mailer/index.js]: Should have each `key: value` pair on a separate line for an object this long.

File: server/mailer/index.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * Sends a password reset successful email.
+ * @param {Object} user The user who has reset their password.
+ * @return {Promise} SendGrid request promise.
+ */
+function sendPasswordResetSuccessfulEmail(user) {
+	return new Promise((resolve, reject) => {
+		if (!SENDGRID_API_KEY) resolve();
+
+		const request = createEmailMsg({ toEmail: user.email, subject: 'QHacks Password Reset Successful', textContent: 'Your password has been updated!', htmlContent: '<p>Your password has been updated!</p>' });
```
Comment (10kc-zack) [inline server/mailer/index.js]: Should have each `key: value` pair on a separate line for an object this long.

File: server/mailer/index.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * Sends a password reset email to the user.
+ * @param {Object} user The user who wishes to reset their password.
+ * @return {Promise} SendGrid request promise.
+ */
+function sendResetPasswordEmail(user) {
+	return new Promise((resolve, reject) => {
+		if (!SENDGRID_API_KEY) resolve();
+
+		const request = createEmailMsg({ toEmail: user.email, subject: 'QHacks Password Reset Email', textContent: `Someone has requested to reset your QHacks account password, click this link to do so: https://${EMAIL_URL_HOST}/update-password/${user.passwordResetHash}`, htmlContent: `<p>Someone has requested to reset your QHacks account password, click this link to do so: http://localhost:8080/update-password/${user.passwordResetHash}</p>` });
```
Comment (10kc-zack) [inline server/mailer/index.js]: Should have each `key: value pair` on a separate line for an object this long.

**Pull request merged**